apis/src/responses/challenge.rs

```diff
@@ -70,5 +72,71 @@ impl ChallengeResponse {
             band_lower: challenge.band_lower,
         })
     }
+
+}
+}
+}
+
+fn is_compatible(challenge: &ChallengeResponse, user: &str, details: &ChallengeDetails) -> bool {
+    details.game_type == GameType::from_str(&challenge.game_type).unwrap()
+        && details.rated == challenge.rated
+        && details.time_mode == challenge.time_mode
+        && details.time_base == challenge.time_base
+        && details.time_increment == challenge.time_increment
+        && match details.color_choice {
+            ColorChoice::Random => challenge.color_choice == ColorChoice::Random,
+            ColorChoice::White => challenge.color_choice == ColorChoice::Black,
+            ColorChoice::Black => challenge.color_choice == ColorChoice::White
+        }
+        && if let Some(opponent) = &details.opponent {
+            opponent == &challenge.challenger.username
+        } else if let Some(opponent) = &challenge.opponent {
+            opponent.username == user
+        } else {
+            challenge.challenger.username != user
+        }
+}
+
+fn has_same_details(challenge: &ChallengeResponse, user: &str, details: &ChallengeDetails) -> bool {
+    let challenge_opponent = challenge
+        .opponent
+        .as_ref()
+        .map(|opponent| opponent.username.as_str());
+
+    details.game_type == GameType::from_str(&challenge.game_type).unwrap()
+        && details.rated == challenge.rated
+        && details.band_lower == challenge.band_upper
+        && details.band_upper == challenge.band_lower
+        && details.time_mode == challenge.time_mode
+        && details.time_base == challenge.time_base
+        && details.time_increment == challenge.time_increment
+        && details.color_choice == challenge.color_choice
+        && challenge_opponent == details.opponent.as_deref()
+        && challenge.challenger.username == user
+}
+
+pub fn create_challenge_handler(
+    user: String,
+    details: ChallengeDetails,
+    challenges: Values<String, ChallengeResponse>,
+) -> Option<ChallengeAction> {
+    let mut matching_id = None;
+    let mut exists = false;
+    for challenge in challenges {
+        if is_compatible(challenge, &user, &details) {
```

Comment:
replace this with a https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find, while what I am suggesting is minimally slower it makes your code a lot more readable, because you don't need to initialize 2 `mut` variables and there is no need for `break`.